ow_lander/scripts/camera_capture_action_server.py

```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python3
+
+# The Notices and Disclaimers for Ocean Worlds Autonomy Testbed for Exploration
+# Research and Simulation can be found in README.md in the root directory of
+# this repository.
+
+import rospy
+
+import actionlib
+
+import ow_lander.msg
+from std_msgs.msg import Empty, Float64
+from sensor_msgs.msg import PointCloud2
+
+class CameraCaptureActionServer(object):
+  #_result = std_msgs.msg.Bool()
+
+  def __init__(self, name):
+    self._action_name = name
+    self._result = ow_lander.msg.CameraCaptureResult()
+    self._exposure_pub = rospy.Publisher(
+      '/gazebo/plugins/camera_sim/exposure', Float64, queue_size=10)
+    self._trigger_pub = rospy.Publisher(
+      '/StereoCamera/left/image_trigger', Empty, queue_size=10)
+    self._point_cloud_sub = rospy.Subscriber('/StereoCamera/points2', PointCloud2, self._handle_point_cloud)
+    self._server = actionlib.SimpleActionServer(self._action_name,
+                                                ow_lander.msg.CameraCaptureAction,
+                                                execute_cb=self.execute_cb,
+                                                auto_start = False)
+    self._server.start()
+
+  def execute_cb(self, goal):
+    # helper variables
+    r = rospy.Rate(1)
+    self._result.point_cloud_created = False
+
+    # Set exposure if it is specified
+    if goal.exposure > 0:
+      self._exposure_pub.publish(goal.exposure)
+      # There is no guarantee that exposure is set before the image is triggered.
+      # Pause to make it highly likely that exposure is received first.
+      r.sleep()
+
+    self._trigger_pub.publish()
+
+    for i in range(0, 5):
+      if self._server.is_preempt_requested():
+        rospy.loginfo(f'{self._action_name}: Preempted')
+        self._server.set_preempted()
+        return
+      if self._result.point_cloud_created:
+        rospy.loginfo(f'{self._action_name}: Succeeded')
+        self._server.set_succeeded(self._result, 'point cloud received')
+        return
+      r.sleep()
+
+    rospy.loginfo(f'{self._action_name}: Failed')
+    self._server.set_aborted(self._result, 'timed out')
+
+  def _handle_point_cloud(self, points):
+    """
+    :type points: sensor_msgs.msg.PointCloud2
+    """
+    
+    # CameraCapture was successful if a point cloud is received. However, there
+    # does not appear to be a way to associate this point cloud with the
+    # original trigger message. A trigger could have been sent using without
+    # using the CameraCapture action client.
```

Comment:
Minor: extra "using" in code comment.  @AstroStucky I don't follow your second sentence.

Reply:
> Minor: extra "using" in code comment. @AstroStucky I don't follow your second sentence.

We can disregard this comment. I did some digging and what I said does not make sense. I thought the `image_trigger` topic calls a function that we created, which is what handles the function calls that make photography happen, but it looks as though that is in fact set up by a thirdparty ROS Package, and `image_trigger` is just the interface it provides.

Reply:
I think this is not an important issue.   Actions are the official interface to autonomy; the camera should be operated only though actions.

Reply:
To elaborate on my original comment, it might be possible to set exposure with a service and block until it is complete. This would require significant work on IRGCameraSimPlugin. (ROS2 allows you to set a parameter on a specific node and get a callback after it is set, which would be a good solution if we were using ROS2.)